src/model/consent/consent.ts

```diff
@@ -0,0 +1,148 @@
+ http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, the Mojaloop
+ files are distributed onan 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ ANY KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ Contributors
+ --------------
+ This is the official list of the Mojaloop project contributors for this file.
+ Names of the original copyright holders (individuals or organizations)
+ should be listed with a '*' in the first column. People who have
+ contributed from an organization can be listed under the organization
+ that actually holds the copyright for their contributions (see the
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Raman Mangla <ramanmangla@google.com>
+ --------------
+ ******/
+
+/*
+ * Unlike MySQL, SQLite (testing DB) doesn't support Timestamp type natively.
+ * It returns ISO strings using the inbuilt Date() function.
+ * Thus, typecasting of MySQL Timestamp to Date object for insertion/retrieval
+ * for 'createdAt' field and testing for the same may be required in future.
+ */
+
+import { NotFoundError } from '../errors'
+import Knex from 'knex'
+
+/*
+ * Interface for Consent resource type
+ */
+export interface Consent {
+  id: string;
+  initiatorId: string;
+  participantId: string;
+  credentialId?: string;
+  credentialType?: string;
+  credentialStatus?: string;
+  credentialPayload?: string;
+  credentialChallenge?: string;
+  createdAt?: Date;
+}
+
+/*
+ * Class to abstract Consent DB operations
+ */
+export class ConsentDB {
+  // Knex instance
+  private Db: Knex
+  // Nullable fields which don't allow for explicit null upsert
+  private static nullProtectedProps: Record<string, boolean> = {
+    credentialId: true,
+    credentialType: true,
+    credentialStatus: true,
+    credentialPayload: true,
+    credentialChallenge: true
+  }
+
+  public constructor (dbInstance: Knex) {
+    this.Db = dbInstance
+  }
+
+  // Add initial Consent parameters
+  public async register (consent: Consent): Promise<number> {
+    // Returns array containing number of inserted rows
+    const insertCount: number[] = await this
+      .Db<Consent>('Consent')
+      .insert(consent)
+
+    return insertCount[0]
+  }
+
+  // Update Consent credential
+  // Only non-null Consent fields are updated
+  public async updateCredentials (consent: Consent): Promise<number> {
+    const validatedConsent = {}
+
+    for (const key in consent) {
+      // Only allow nullProtectedFields with non-null values for update
+      if (ConsentDB.nullProtectedProps[key] && (consent as unknown as Record<string, boolean>)[key]) {
```

Comment:
Completely out of my depth, but what about:

`consent[key as keyof typeof consent]`

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html

?

Also, we should also check for explicit nulls or undefined. Because 0 values (not null, but 0 integers) would throw an error here when they shouldn't! Or look at hapi joi for validation? Or maybe just don't do this at all here and leave this for the handler layer to address?